effect/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -0,0 +1,547 @@
+  }
+
+  /**
+   * Recovers from some or all of the error cases.
+   *
+   * {{{
+   * openFile("data.json").catchSome {
+   *   case FileNotFoundException(_) => openFile("backup.json")
+   * }
+   * }}}
+   */
+  final def catchSome(pf: PartialFunction[Throwable, IO[A]]): IO[A] = {
+    def tryRescue(t: Throwable): IO[A] = t match {
+      case t if pf.isDefinedAt(t) => pf(t)
+      case t => IO.fail(t)
+    }
+
+    self.attempt.flatMap(_.fold(tryRescue)(IO.now(_)))
+  }
+
+  /**
+   * Maps this action to the specified constant while preserving the
+   * effects of this action.
+   */
+  final def const[B](b: => B): IO[B] = self.map(_ => b)
+
+  /**
+   * A variant of `flatMap` that ignores the value produced by this action.
+   */
+  final def *> [B](io: => IO[B]): IO[B] = self.flatMap(_ => io)
+
+  /**
+   * Sequences the specified action after this action, but ignores the
+   * value produced by the action.
+   */
+  final def <* [B](io: => IO[B]): IO[A] = self.flatMap(io.const(_))
+
+  /**
+   * Repeats this action forever (until the first error).
+   */
+  final def forever[B]: IO[B] = self *> self.forever
+
+  /**
+   * Retries continuously until this action succeeds.
+   */
+  final def retry: IO[A] = self orElse retry
+
+  /**
+   * Retries this action the specified number of times, until the first success.
+   * Note that the action will always be run at least once, even if `n < 1`.
+   */
+  final def retryN(n: Int): IO[A] =
+    if (n <= 1) self
+    else self orElse (retryN(n - 1))
+
+  /**
+   * Retries continuously until the action succeeds or the specified duration
+   * elapses.
+   */
+  final def retryFor(duration: Duration): IO[A] =
+    IO.absolve(
+      retry.attempt race (IO.sleep(duration) *>
+                          IO(-\/(Errors.TimeoutException(duration)))))
+
+  /**
+   * Maps this action to one producing unit, but preserving the effects of
+   * this action.
+   */
+  final def toUnit: IO[Unit] = const(())
+
+  /**
+   * Calls the provided function with the result of this action, and
+   * sequences the resulting action after this action, but ignores the
+   * value produced by the action.
+   *
+   * {{{
+   * for {
+   *   file <- readFile("data.json").peek(putStrLn(_))
+   * } yield file
+   * }}}
+   */
+  final def peek[B](f: A => IO[B]): IO[A] = self.flatMap(a => f(a).const(a))
+
+  /**
+   * Times out this action by the specified duration.
+   *
+   * {{{
+   * action.timeout(1.second)
+   * }}}
+   */
+  final def timeout(duration: Duration): IO[A] = {
+    val err: IO[Throwable \/ A] = IO(-\/(Errors.TimeoutException(duration)))
+
+    IO.absolve(self.attempt.race(err.delay(duration)))
+  }
+
+  /**
+   * Delays this action by the specified amount of time.
+   */
+  final def delay(duration: Duration): IO[A] =
+    IO.sleep(duration) *> self
+
+  /**
+   * An integer that identifies the term in the `IO` sum type to which this
+   * instance belongs (e.g. `IO.Tags.Point`).
+   */
+  def tag: Int
```

Comment:
Should this be public?

Reply:
I could go either way on this. There's no unsafety in this API, but there's also no utility in the API unless you're going to write your own interpreter.